### DISCUSSION TITLE
How to implement something like "pop if possible, else go"

### DISCUSSION THREAD

**Discussion by bw-flagship** [csells/go_router#333] [Q&A]
``I have a scenario where I struggle to find the correct solution with go_router.

Imagine I have a BookListView (with a state, e.g. my scroll position) and a BookDetailView with the bookId as a route parameter.
The BookDetailView contains a back-button. When I press it, I want to navigate to the BookListView.
When comming from the BookListView, I archive this via pop().
However, when I access the BookDetailView via deeplink, I need to use context.go(…), because the BookListView does not exist in the stack.
I can't always use context.go because I would lose the state of my BookListView.

I would need something like:
```
if (currentPagesCount > 1){
    pop();
} else {
    context.go("book-list-view");
}
```

Right now I solve it like this:
```
try {
      router.pop();
    } catch (_) {
      router.goNamed(
        RouteNames.home,
        params: {
          "tab": HomeViewTab.airspaces.name,
        },
      );
    }

```

But it feels very wrong. What would be the better way?

Thanks alot!

**Comment by csells** [comment]
You can get the current stack size with something like: `GoRouter.of(context).routerDelegate.matches.length`.

**Reply by bw-flagship** [reply]
Thanks! That works, but produces a warning because the "matches" field is marked as "visible for testing only".

I think it would be a good idea to remove the "visibleForTesting"-Attribute, since it is useful to get the matches.
Is there anything wrong with that? If not, I could submit a PR.

![2022-02-14_18-33-00](https://user-images.githubusercontent.com/89379862/153916356-80f72a72-0087-4ac7-91fc-15c4556733a6.png)
![2022-02-14_18-33-59](https://user-images.githubusercontent.com/89379862/153916365-2829450e-1dea-4863-97a5-d7760ac85505.png)
